packages/cozy-stack-client/src/getIconURL.js

```diff
@@ -0,0 +1,126 @@
+import memoize from './memoize'
+
+const mimeTypes = {
+  gif: 'image/gif',
+  ico: 'image/vnd.microsoft.icon',
+  jpeg: 'image/jpeg',
+  jpg: 'image/jpeg',
+  png: 'image/png',
+  svg: 'image/svg+xml'
+}
+
+const fetchOptions = stackClient => {
+  return {
+    headers: {
+      Authorization: 'Bearer ' + stackClient.token
+    },
+    credentials: true
+  }
+}
+
+const getIconExtensionFromApp = app => {
+  if (!app.icon) {
+    throw new Error(
+      `${app.name}: Cannot detect icon mime type since app has no icon`
+    )
+  }
+
+  const extension = app.icon.split('.').pop()
+
+  if (!extension) {
+    throw new Error(
+      `${app.name}: Unable to detect icon mime type from extension (${
+        app.icon
+      })`
+    )
+  }
+
+  return extension
+}
+
+const fallbacks = async (tries, check) => {
+  let err
+  for (let _try of tries) {
+    try {
+      const res = await _try()
+      check && check(res)
+      return res
+    } catch (e) {
+      err = e
+    }
+  }
+  throw err
+}
+
+const _getIconURL = async (stackClient, type, slug, appData) => {
+  const fetchOpts = fetchOptions(stackClient)
+  const resp = await fallbacks(
```

Comment:
If I understand well here we are always trying to get the icon from an installed app first. It means in the Cozy-Store, that we will have more than 100 404 errors at every load.

As we have `appData` as a parameter and as thanks to it we are able to differenciate installed apps from registry, I think we should use `appData` if it is passed.

Another way could be to fetch first registry icons, but the icons can differ between registry and stack (when a newer registry version embed a different icon for example)

Reply:
> e are able to differenciate installed apps from registry, I think we should use `appData` if it is passed.

How do you differentiate between registry app and installed app ?

Reply:
I added a priority option (so I changed to a option parameters) so that it is possible to indicate that you prefer the registry first.

Reply:
I will not add this to AppCollection at this point. I find that instantiating a Collection to get this method is a bit cumbersome.